Gutenberg: Load Jetpack blocks in Calypso

dmsnell opened the pull request:
Supersedes #26692 
Resolves #27634

As we've been building our Jetpack collection of Gutenberg extensions we
have been primarily doing so through self-hosted Jetpack test
environments.

With this change we'll start loading the Jetpack blocks inside Calypso
(where we plan on having them load anyway long-term) and that should
help establish a tighter feedback loop on parts of that block and
extension development.

![jetpack-blocks mov](https://user-images.githubusercontent.com/5431237/46985642-15e63f80-d0b1-11e8-9aa6-eff815876b8e.gif)

**Status**

 - Seems to be working and properly loading the Jetpack blocks.
 - `calypso.live` is failing to build, failing to load the babel loader, not sure why
 -  getting `Jed localization error: Error: Domain `jetpack` was not found.` in the console

**Testing**

Visit `/gutenberg/` and create a new post on some Simple Site.
Add blocks: the existing Jetpack blocks are the markdown block,
tiled gallery, and related posts.

Jetpack should also exist in the block inserter as a category.

dmsnell committed:
Gutenberg: Load Jetpack blocks in Calypso

As we've been building our Jetpack collection of Gutenberg extensions we
have been primarily doing so through self-hosted Jetpack test
environments.

With this change we'll start loading the Jetpack blocks inside Calypso
(where we plan on having them load anyway long-term) and that should
help establish a tighter feedback loop on parts of that block and
extension development.

dmsnell commented:
I can't seem to get these things to load due to the missing `React` import.

```
Uncaught SyntaxError: Unexpected token <
```

I've been playing around with the `wepback` config for `babel` and trying to use `plugin-transform-react-jsx` and also the [loader provided by core](https://github.com/WordPress/gutenberg/tree/master/packages/babel-plugin-import-jsx-pragma) and none seem to be working for me.

this has been super helpful in understanding how to load plugins for a specific path: https://babeljs.io/docs/en/options

at times when I set a `test` value in the babel config as `/gutenberg\/extensions\//` it tells me `.overrides[0].test must be a string/Function/RegExp, or an array of those` and yet obviously it is, even with `new RegExp()` syntax it fails - who knows why this is failing

maybe some wizards like @sirreal or @jsnajdr can help us figure out how to get this plugin loading in the gutenberg/extensions directory

dmsnell committed:
update to make things work

dmsnell commented on webpack.config.js:
```diff
@@ -168,7 +158,36 @@ function getWebpackConfig( { cssFilename, externalizeWordPressPackages = false }
+								],
+							},
+						},
 					],
```
falling down the cliff 😄

dmsnell commented on webpack.config.js:
```diff
@@ -168,7 +158,36 @@ function getWebpackConfig( { cssFilename, externalizeWordPressPackages = false }
+													pragma: 'createElement',
+												},
+											],
+										],
```
@sirreal and @ockham this is where I was thinking we might hook in via an argument to `getBaseConfig()`.

the `overrides` in Babel seem quite useful in any case

dmsnell commented:
seems that my original troubles were masked by a separate and independent issue. the blocks weren't appearing but that was probably a timing issue with loading the editor. by moving the import statement into the editor React component we are able to make sure that the blocks register after the editor instance has loaded. before then and they won't load because the editor either doesn't exist or disappears.

dmsnell committed:
move import to dependencies from dev deps

dmsnell committed:
fix babel plugin name

jsnajdr commented on webpack.config.js:
```diff
@@ -168,7 +158,36 @@ function getWebpackConfig( { cssFilename, externalizeWordPressPackages = false }
+								babelrc: false,
+								cacheDirectory: path.join( __dirname, 'build', '.babel-client-cache' ),
+								cacheIdentifier,
+								overrides: [
```
Just driving by, seeing that the `test` validation issue has been already solved (?), but I'm wondering: could the `overrides` field be in the `babel.config.js` rather than the webpack config? Other users of Babel in the repo (namely Jest tests) also want the `client/gutenberg/extensions` stuff to be correctly transpiled, don't they?

ockham commented:
>  getting `Jed localization error: Error: Domain `jetpack` was not found.` in the console

This is going to be interesting. /cc @Automattic/i18n for advice -- we've started to put JP Gutenblock translations into that textdomain. How do we best make them available in Calypso?

tyxla commented:
Hm, perhaps we could think about introducing configuration settings to the SDK, on a per-preset basis, and have the textdomain as one of them? If we have that, when building the presets, we could have the flexibility to set different textdomains for the different presets. We should only make sure we ship all the translations from .com (which I'm not sure if we're not already doing).

dmsnell committed:
update shrinkwrap

tyxla commented:
> @tyxla can you reword this? do you mean the way we are passing options into getBaseConfig()?

I didn't mean a particular approach, but that would be one possible way to do it. We could add a new argument to the SDK, or implement some configs in the presets and pull them from there. Plenty of options to explore 😉

sirreal committed:
Move markdown-it from devDep to normal dependencies

This is now a production asset, Docker builds fail because Docker
installs production assets only.
